cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -188,6 +188,79 @@ Result<compute::Declaration> FromProto(const substrait::Rel& rel,
       });
     }
 
+    case substrait::Rel::RelTypeCase::kJoin: {
+      const auto& join = rel.join();
+      RETURN_NOT_OK(CheckRelCommon(join));
+
+      if (!join.has_left()) {
+        return Status::Invalid("substrait::JoinRel with no left relation");
+      }
+
+      if (!join.has_right()) {
+        return Status::Invalid("substrait::JoinRel with no right relation");
+      }
+
+      compute::JoinType join_type;
+      switch (join.type()) {
+        case 0:
+          return Status::NotImplemented("Unspecified join type is not supported");
+        case 1:
+          join_type = compute::JoinType::INNER;
+          break;
+        case 2:
+          return Status::NotImplemented("Outer join type is not supported");
+        case 3:
+          return Status::NotImplemented("Left join type is not supported");
+        case 4:
+          return Status::NotImplemented("Right join type is not supported");
+        case 5:
+          return Status::NotImplemented("Semi join type is not supported");
+        case 6:
+          return Status::NotImplemented("Anti join type is not supported");
+        default:
+          return Status::Invalid("Unsupported join type");
+      }
+
+      ARROW_ASSIGN_OR_RAISE(auto left, FromProto(join.left(), ext_set));
+      ARROW_ASSIGN_OR_RAISE(auto right, FromProto(join.right(), ext_set));
+
+      if (!join.has_expression()) {
+        return Status::Invalid("substrait::JoinRel with no expression");
+      }
+
+      ARROW_ASSIGN_OR_RAISE(auto expression, FromProto(join.expression(), ext_set));
+
+      const auto& callptr = expression.call();
```

Comment:
Is this a pointer?  If so I think we can use `const auto*` (but I may be wrong on this)

Reply:
This is how we arrive at this >> 

`const Call* call() const;` 

I see your point. But with `auto`, the existing usage in the code base is like this

```c++
auto call = expr.call();
```

I might have overworked this :/ WDYT?

Reply:
`auto` is very odd with pointers:

```
int x = 7;
auto auto_only = &x; // auto_only will have type: int*
auto* auto_ptr = &x; // auto_ptr will have type: int*
const autu& auto_cref = &x; // auto_cref will have type: int * const &
```

General rule of thumb is that `auto` is ok but `auto*` is more explicit (`auto* auto_ptr = x` will fail to compile) and also helpful to the reader so they know what is being returned.  We don't really follow any strict pattern within the Arrow code base today so I think both `auto` and `auto*` would be acceptable in this case.  However, `const auto&` is doing something different and should not be used here.

[Example](https://godbolt.org/z/b4KnKfTM6)

Reply:
Thanks, @westonpace good point. I will update.